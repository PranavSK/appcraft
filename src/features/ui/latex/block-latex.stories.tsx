import type { Meta, StoryObj } from '@storybook/react';

import { BlockLatex } from './latex';

const meta: Meta<typeof BlockLatex> = {
  component: BlockLatex,
  decorators: [
    (Story) => (
      <p className="bg-gray-100 p-4">
        The following story <Story /> is wrapped in a paragraph with a gray background to show the
        difference between the inline and block modes.
      </p>
    ),
  ],
};
export default meta;
type Story = StoryObj<typeof BlockLatex>;

export const Block = {
  args: {
    latex: '\\sqrt{2}',
  },
} satisfies Story;
