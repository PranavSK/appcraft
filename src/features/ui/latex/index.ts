export { BlockLatex, InlineLatex } from './latex';
