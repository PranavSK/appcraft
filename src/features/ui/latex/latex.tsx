import katex from 'katex';
import { forwardRef, useMemo } from 'react';

import { cn } from '#/lib/utils';

import type { LatexProps } from './latex.types';

export const Latex = forwardRef<HTMLDivElement, LatexProps>(
  ({ latex, displayMode = false, className, ...props }, ref) => {
    const html = useMemo(
      () =>
        katex.renderToString(latex, {
          displayMode: displayMode,
          errorColor: '#ef4444',
          macros: {},
          throwOnError: false,
        }),
      [displayMode, latex],
    );
    return (
      <div
        data-testid="latex"
        className={cn(!displayMode && 'inline', className)}
        dangerouslySetInnerHTML={{ __html: html }}
        ref={ref}
        {...props}
      />
    );
  },
);
Latex.displayName = 'Latex';
