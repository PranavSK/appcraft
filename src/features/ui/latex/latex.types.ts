import { type HTMLAttributes } from 'react';

export interface LatexProps extends Omit<HTMLAttributes<HTMLDivElement>, 'children'> {
  latex: string;
  displayMode?: boolean;
}
