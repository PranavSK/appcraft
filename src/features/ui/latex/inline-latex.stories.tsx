import type { Meta, StoryObj } from '@storybook/react';

import { InlineLatex } from './latex';

const meta: Meta<typeof InlineLatex> = {
  component: InlineLatex,
  decorators: [
    (Story) => (
      <p className="bg-gray-100 p-4">
        The following story <Story /> is wrapped in a paragraph with a gray background to show the
        difference between the inline and block modes.
      </p>
    ),
  ],
};
export default meta;
type Story = StoryObj<typeof InlineLatex>;

export const Inline = {
  args: {
    latex: '\\sqrt{2}',
  },
} satisfies Story;
