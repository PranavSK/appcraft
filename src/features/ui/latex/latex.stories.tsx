import type { Meta, StoryObj } from '@storybook/react';

import { Latex } from './latex';

const meta: Meta<typeof Latex> = {
  component: Latex,
  decorators: [
    (Story) => (
      <p className="bg-gray-100 p-4">
        The following story <Story /> is wrapped in a paragraph with a gray background to show the
        difference between the inline and block modes.
      </p>
    ),
  ],
};
export default meta;
type Story = StoryObj<typeof Latex>;

export const Inline = {
  args: {
    latex: '\\sqrt{2}',
    displayMode: false,
  },
} satisfies Story;

export const Block = {
  args: {
    latex: '\\sqrt{2}',
    displayMode: true,
  },
} satisfies Story;
