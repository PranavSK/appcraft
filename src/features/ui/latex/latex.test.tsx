import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import { Latex } from './latex';

describe('Latex', () => {
  it('should render html displaying the latex', () => {
    render(<Latex latex="\\sqrt{2}" />);
    expect(screen.getByTestId('latex')).toMatchSnapshot();
  });
});
