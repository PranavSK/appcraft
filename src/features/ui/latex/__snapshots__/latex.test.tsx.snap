// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Latex > should render html displaying the latex 1`] = `
<div
  class="inline"
  data-testid="latex"
>
  <span
    class="katex"
  >
    <span
      class="katex-mathml"
    >
      <math
        xmlns="http://www.w3.org/1998/Math/MathML"
      >
        <semantics>
          <mrow>
            <mspace
              linebreak="newline"
            />
            <mi>
              s
            </mi>
            <mi>
              q
            </mi>
            <mi>
              r
            </mi>
            <mi>
              t
            </mi>
            <mn>
              2
            </mn>
          </mrow>
          <annotation
            encoding="application/x-tex"
          >
            \\\\sqrt{2}
          </annotation>
        </semantics>
      </math>
    </span>
    <span
      aria-hidden="true"
      class="katex-html"
    >
      <span
        class="mspace newline"
      />
      <span
        class="base"
      >
        <span
          class="strut"
          style="height:0.8389em;vertical-align:-0.1944em;"
        />
        <span
          class="mord mathnormal"
        >
          s
        </span>
        <span
          class="mord mathnormal"
          style="margin-right:0.03588em;"
        >
          q
        </span>
        <span
          class="mord mathnormal"
          style="margin-right:0.02778em;"
        >
          r
        </span>
        <span
          class="mord mathnormal"
        >
          t
        </span>
        <span
          class="mord"
        >
          <span
            class="mord"
          >
            2
          </span>
        </span>
      </span>
    </span>
  </span>
</div>
`;
